lib/jxl/enc_color_management.cc

```diff
@@ -887,11 +887,31 @@ Status ColorEncoding::SetFieldsFromICC() {
   JXL_RETURN_IF_ERROR(skcms_Parse(icc_.data(), icc_.size(), &profile));
 
   // skcms does not return the rendering intent, so get it from the file. It
-  // is encoded as big-endian 32-bit integer in bytes 60..63.
+  // should be encoded as big-endian 32-bit integer in bytes 60..63.
   uint32_t rendering_intent32 = icc_[67];
-  if (rendering_intent32 > 3 || icc_[64] != 0 || icc_[65] != 0 ||
-      icc_[66] != 0) {
-    return JXL_FAILURE("Invalid rendering intent %u\n", rendering_intent32);
+  if (icc_[64] == 0 && icc_[65] == 0 && icc_[66] == 0) {
+    // This is the big-endian as the ICC spec requires it case.
+    if (rendering_intent32 > 3) {
+      return JXL_FAILURE("Invalid rendering intent %u\n", rendering_intent32);
+    }
+  } else {
+    // This case has two subcases: the rendering intent
+    // - is completely invalid -> we fail or
+    // - is invalid because is it provided as little-endian ->  we warn.
+    if ((icc_[67] != 0) || icc_[65] != 0 || icc_[66] != 0 || icc_[64] > 3) {
+      return JXL_FAILURE(
+          "Invalid rendering intent bytes: [0x%02X 0x%02X 0x%02X 0x%02X]\n",
+          icc_[64], icc_[65], icc_[66], icc_[67]);
+    }
+    rendering_intent32 = icc_[64];
+    JXL_WARNING(
+        "Invalid rendering intent bytes: [0x%02X 0x%02X 0x%02X 0x%02X], "
+        "assuming %u was meant",
+        icc_[64], icc_[65], icc_[66], icc_[67], rendering_intent32);
+  }
+  uint32_t combined_rending_intent = rendering_intent32 + (icc_[66] << 8) +
+                                     (icc_[65] << 16) + (icc_[64] << 24);
+  if (combined_rending_intent > 3) {
```

Comment:
This is empty. Perhaps the similar check above could be moved here? Also, in the big endian case, `combined_rendering_intent` ends up with `icc_[64] + icc_[64] << 24` as far as I can tell.

Perhaps clearer would be to remove all this and just:

1. compute `big_endian_rendering_intent` from all four bytes (no “just take the last byte and check that the other bytes are zero” trick);
2. compute `little_endian_rendering_intent` from all four bytes in the other direction;
3. take the min and warn if that min is the little-endian version;
4. return an error if it’s more than 3.

What do you think?

Reply:
I like that approach, will do it like this!